Python/Product/VSInterpreters/CondaEnvironmentFactoryProvider.cs

```diff
@@ -0,0 +1,372 @@
+// Python Tools for Visual Studio
+// Copyright(c) Microsoft Corporation
+// All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the License); you may not use
+// this file except in compliance with the License. You may obtain a copy of the
+// License at http://www.apache.org/licenses/LICENSE-2.0
+//
+// THIS CODE IS PROVIDED ON AN  *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS
+// OF ANY KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY
+// IMPLIED WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
+// MERCHANTABLITY OR NON-INFRINGEMENT.
+//
+// See the Apache Version 2.0 License for specific language governing
+// permissions and limitations under the License.
+
+using System;
+using System.Collections.Generic;
+using System.ComponentModel.Composition;
+using System.Diagnostics;
+using System.IO;
+using System.Linq;
+using System.Threading;
+using Microsoft.PythonTools.Infrastructure;
+using Newtonsoft.Json;
+
+namespace Microsoft.PythonTools.Interpreter {
+    /// <summary>
+    /// Detects interpreters in user-created conda environments.
+    /// </summary>
+    /// <remarks>
+    /// Uses %HOMEPATH%/.conda/environments.txt and `conda info --envs`.
+    /// </remarks>
+    [InterpreterFactoryId(FactoryProviderName)]
+    [Export(typeof(IPythonInterpreterFactoryProvider))]
+    [Export(typeof(CondaEnvironmentFactoryProvider))]
+    [PartCreationPolicy(CreationPolicy.Shared)]
+    class CondaEnvironmentFactoryProvider : IPythonInterpreterFactoryProvider {
+        private readonly Dictionary<string, PythonInterpreterInformation> _factories = new Dictionary<string, PythonInterpreterInformation>();
+        internal const string FactoryProviderName = "CondaEnv";
+        internal const string EnvironmentCompanyName = "CondaEnv";
+        private int _ignoreNotifications;
+        private bool _initialized;
+        private readonly CPythonInterpreterFactoryProvider _globalProvider;
+        private readonly bool _watchFileSystem;
+        private FileSystemWatcher _envsTxtWatcher;
+        private Timer _envsTxtWatcherTimer;
+        private string _environmentsTxtPath;
+
+        [ImportingConstructor]
+        public CondaEnvironmentFactoryProvider(
+            [Import] CPythonInterpreterFactoryProvider globalProvider,
+            [Import("Microsoft.VisualStudioTools.MockVsTests.IsMockVs", AllowDefault = true)] object isMockVs = null
+        ) {
+            _watchFileSystem = isMockVs == null;
+            _globalProvider = globalProvider;
+        }
+
+        private void EnsureInitialized() {
+            lock (this) {
+                if (!_initialized) {
+                    _initialized = true;
+                    _environmentsTxtPath = Path.Combine(
+                        Environment.GetFolderPath(Environment.SpecialFolder.UserProfile),
+                        ".conda",
+                        "environments.txt"
+                    );
```

Comment:
Historically, we've found this kind of code to be pretty fragile. Should handle any errors that occur and quietly ignore the environments.txt file if we can't get a path.